public/editor/scripts/editor/js/fc/bramble-ui-bridge.js

```diff
@@ -11,10 +11,72 @@ define(function(require) {
 
   var _escKeyHandler;
 
+  var adapting = false;
+  var adaptTimeoutMS = 200; // How often we adapt editor bar layout
+
   function updateLayout(data) {
     $(".filetree-pane-nav").width(data.sidebarWidth);
     $(".editor-pane-nav").width(data.firstPaneWidth);
     $(".preview-pane-nav").width(data.secondPaneWidth);
+
+    // Only adapt the layout every once in a while
+    if(!adapting) {
+      adapting = true;
+      adaptLayout();
+      setTimeout(function(){
+        adapting = false;
+      },adaptTimeoutMS);
+    }
+  }
+
+  // Adapt each of the pane header elements
+  function adaptLayout(){
+    $(".nav-container").each(function(){
+      adaptElement($(this));
+    });
+  }
+
+  // Checks if there is enough room for all of the elements inside it
+  // Adds a 'narrow' class, in priority order, when there isn't.
+  function adaptElement(el){
+    var itemCount = el.find("[adapt-order]").length;
+    el.find("[adapt-order]").addClass("narrow");
+
+    for(var i = itemCount; i > 0; i--) {
+      if(hasEnoughRoom(el)) {
+        el.find("[adapt-order="+i+"]").removeClass("narrow");
+      } else {
+        break;
+      }
+      if(!hasEnoughRoom(el)) {
+        el.find("[adapt-order="+i+"]").addClass("narrow");
+      }
+    }
+  }
+
+  // Checks if the current element has enough room for everything in it
+  // by checking if the last visible element is too far to the right.
+  function hasEnoughRoom(el) {
+    var maxRight = el[0].getBoundingClientRect().width - parseInt(el.css("padding-left"));
+
+    // Finds the last visible first-order child
+    var lastEl = false;
+    el.find("> *").each(function(){
+      if($(this).is(":visible")){
+        lastEl = $(this);
+      }
+    });
+
+    if(lastEl) {
+      var parentLeft = el[0].getBoundingClientRect().left;
```

Comment:
The parent alement isnt the same as the "el" element here, so we do need to get boundingClientRect twice